### PR TITLE
New Adwaita View-Switcher Example

### PR DESCRIPTION
**Pull request by tronta**
I'm trying to create a new example for using a view switcher in adwaita.

### CODE REVIEW

File: relm4-examples/libadwaita/examples/view-switcher.rs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+                append = &adw::HeaderBar {
+                    set_title_widget = Some(&adw::ViewSwitcherTitle) {
+                        set_title: "A stack switcher App",
+                        set_stack: Some(&stack),
+                    }
+                },
+                
+                append: stack = &adw::ViewStack {
+                    add_titled: args!(
+                        &gtk::Label::new(Some("This is the start page")), Some("First"), "First Page"),
```
Comment (tronta) [inline relm4-examples/libadwaita/examples/view-switcher.rs:60]: Maybe you could help me on that:
I want to create those subpages, which use multiple arguments. Now I want to continue below with the macro notation. Do need to use components or such? Or is there something I am missing so far?

Comment (AaronErhardt) [reply relm4-examples/libadwaita/examples/view-switcher.rs:60]: @tronta you should be able to use `add_titled(arg1, arg2, ...) = &gtk::Label ... `

Comment (tronta) [reply relm4-examples/libadwaita/examples/view-switcher.rs:60]: The first parameter of the function is expecting the label, so I'm not sure how I can assign it as you proposed.

Comment (tronta) [reply relm4-examples/libadwaita/examples/view-switcher.rs:60]: Indeed, it's working like this. You have probably described it somewhere, but I was not expecting this behavior. Maybe I forgot this. Thanks

File: relm4-examples/libadwaita/examples/view-switcher.rs [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+                append = &adw::HeaderBar {
+                    set_title_widget = Some(&adw::ViewSwitcherTitle) {
+                        set_title: "A stack switcher App",
+                        set_stack: Some(&stack),
+                    }
+                },
+                
+                append: stack = &adw::ViewStack {
+                    add_titled(Some("First"), "First Page") = &gtk::Label {
+                        set_label: "This is the start page",
```
Comment (tronta) [inline relm4-examples/libadwaita/examples/view-switcher.rs:60]: Sorry, I'm stuck. What would be the way to add icons and such?
To me the way how to untilize [set_icon_name](https://world.pages.gitlab.gnome.org/Rust/libadwaita-rs/stable/latest/docs/libadwaita/struct.ViewStackPage.html#method.set_icon_name) is still mysterious

Comment (AaronErhardt) [reply relm4-examples/libadwaita/examples/view-switcher.rs:60]: Before using set_icon_name you either have to know how your theme calls a certain icon (which is not very reliable) or you add your own image as resource (there's an "Icon library" app to help you btw.). Using your own resources is a bit painful though because it usually requires meson as build system and a resource schema to load the image. It think you can also use resources with cargo and pure Rust but I haven't tried it yet.

An example code to load a resource files can be found here: https://gitlab.elektrotechnik.hs-augsburg.de/erhardt/state-factory/-/blob/gtk4/src/static_resources.rs (lines 22-24)

And together with this function you might be able to get images working without meson: https://gtk-rs.org/gtk-rs-core/git/docs/gio/fn.compile_resources.html

My idea would be: compile your XML schema (schemas look like [this](https://gitlab.elektrotechnik.hs-augsburg.de/erhardt/state-factory/-/blob/gtk4/data/resources/resources.gresource.xml)) -> load compiled schema -> register schema for your application -> use icon

Comment (tronta) [reply relm4-examples/libadwaita/examples/view-switcher.rs:60]: Probably I should have been more precise with my problem description. I understand the concept with the named icons. The problem I have here is, how to create a `libadwaita::ViewStackPage` as it does not seam to have a constructor and somehow just gets implicitly created by using e.g. `add_titled`. But how can I later add  those icons to it?

Comment (tronta) [reply relm4-examples/libadwaita/examples/view-switcher.rs:60]: An other question I have: Can I also combine `property_name(value1, value2, ...) = gtk::Box { ... }` with providing a name `property_name: name = gtk::Box { ... }`?
Maybe something like `property_name: name(value1, value2, ...) = gtk::Box { ... }`

Comment (AaronErhardt) [reply relm4-examples/libadwaita/examples/view-switcher.rs:60]: The problem is that Relm4 doesn't handle cases yet where a property function returns a new widget. Libadwaita differs from GTK in this point and the macro will probably need a new syntax for this.

An yes you should be able to use a name for the widget. I think the syntax is like this:
`property(arg1, arg2, ...): name = gtk::Box { ... }`

Comment (tronta) [reply relm4-examples/libadwaita/examples/view-switcher.rs:60]: Would it make sense to make them aware of this problem?
Maybe they could just add additional methods to avoid that problem. Or do you think it's not worth the effort?

Comment (AaronErhardt) [reply relm4-examples/libadwaita/examples/view-switcher.rs:60]: I don't think they would introduce a new API for us unless it males sense in general. So if you have an idea for a new ViewStack API an issue would make sense. Otherwise Relm4 needs to adapt, I guess.

Comment (tronta) [reply relm4-examples/libadwaita/examples/view-switcher.rs:60]: I would think about something like this
``` rust
                append: stack = &adw::ViewStack {
                    add: &adw::ViewStackPage {
                        set_child: = &gtk::Label {
                            set_label: "This is the start page",
                        },
                        set_icon_name: Some("icon_name"),
                        set_badge_number: 3,
                    },
                    add_titled(Some("First"), "First Page") = &gtk::Label {
                        set_label: "This is the start page",
                    },
```
What do you think about that? This would probably need two additional methods:

- `adw::ViewStackPage::new()`
- `adw::ViewStackPage::set_child(&self, child: &P)`

Comment (AaronErhardt) [reply relm4-examples/libadwaita/examples/view-switcher.rs:60]: That's what I'd suggest as well. We need some sort of constructor and a method to set the child. Or even both at once. Do you have time to open an issue over at libadwaita?

Comment (tronta) [reply relm4-examples/libadwaita/examples/view-switcher.rs:60]: Yes, currently I have time, but not sure, how the C-API looks like :smile: 
I'll will have a look.

Comment (AaronErhardt) [reply relm4-examples/libadwaita/examples/view-switcher.rs:60]: You're basically looking for a new `view_stack_page_new` and `view_stack_page_set_child`method when "talking C".
More here: https://gnome.pages.gitlab.gnome.org/libadwaita/doc/1.0.0.alpha.4/

Comment (tronta) [reply relm4-examples/libadwaita/examples/view-switcher.rs:60]: What do you think? https://gitlab.gnome.org/GNOME/libadwaita/-/issues/311

Comment (AaronErhardt) [reply relm4-examples/libadwaita/examples/view-switcher.rs:60]: Looks good. Let's see what they think :)

Comment (AaronErhardt) [reply relm4-examples/libadwaita/examples/view-switcher.rs:60]: Thanks for opening the issue! I already expected they wouldn't be happy about these changes (and to be honest I wouldn't be either in their place). That then means more work for me and a new syntax for handling these types of functions... suggestions are welcome :)

**Comment by tronta** [comment]
What do you think about something like this?
``` Rust
                append: stack = &adw::ViewStack {
                    add_titled(Some("First"), "First Page") = &gtk::Label {
                        set_label: "This is the start page",
                    },
                    add_titled(Some("Second"), "Second Page") = &gtk::Label {
                        set_label: "This is the second page"
                    }.unwrap().set_badge_number(3),
                    v1 = add_titled(Some("Third"), "Third Page") = &gtk::Label {
                        set_label: "This is the last page"
                    }.unwrap().set_badge_number(3)
                }
```

**Comment by AaronErhardt** [comment]
Mhh, maybe something like this? Using `returns` as a keyword to use the returned value and `?` to unwrap it.
```rust
                append: stack = &adw::ViewStack {
                    add_titled(Some("First"), "First Page") = &gtk::Label {
                        set_label: "This is the start page",
                    },
                    add_titled(Some("Second"), "Second Page") = &gtk::Label {
                        returns? = page: ViewStackPage {
                            set_badge_number: 3,
                        },
                        set_label: "This is the second page"
                    },
                }
```

**Comment by tronta** [comment]
tbh I wouldn't expect the return value probably not inside the child definition, but its feasibility is of coarse important :smile: 

What about this?
```rust
                append: stack = &adw::ViewStack {
                    add_titled(Some("First"), "First Page") = &gtk::Label {
                        set_label: "This is the start page",
                    } -> page: ViewStackPage? {
                        set_badge_number: 3,
                    },
                    add_titled(Some("Second"), "Second Page") = &gtk::Label {
                        set_label: "This is the second page"
                    }
                }
```

**Comment by tronta** [comment]
Great. How much effort is something like this? I have no clue. Haven't touched macro programming yet :disappointed:

**Comment by AaronErhardt** [comment]
I don't know honestly. It's sometimes hard to tell when it comes to macros. In any case, it will probably take until the weekend before I'll be able to send a patch for this.

**Comment by tronta** [comment]
Don't worry. Thanks a lot for you effort!

**Comment by AaronErhardt** [comment]
Btw. I had a look at the macro. There might be a hack to implement this feature faster but I think it's time to re-organize the macro logic into a more understandable and extensible structure while adding this feature. 
That makes even more sense with micro components being added that require a new widget macro and I don't want to copy paste a lot of code.

**Comment by tronta** [comment]
A solid code base sounds like a great idea!

**Comment by AaronErhardt** [comment]
As an update on this matter: I've re-organized the macro structure to be much more clear now and able of handling the new syntax (actually parsing already works, but code generation is still missing). 

The problem is however, that the structure of Relm4 currently makes it impossible to handle this conveniently. Currently, both widgets and models are initialized from the root (the app) to the leaves (the outermost components). This is problematic because to keep the order of widgets and components in sync with the order they appear in the macro, the all widgets must be connected when the components are ready, which is after the widgets are actually created. Therefore the [connect_components method](https://aaronerhardt.github.io/docs/relm4/relm4/trait.Widgets.html#method.connect_components) exists to connect the components to the parents widgets in an extra step and is used by the macro to ensure correct ordering (see https://github.com/AaronErhardt/relm4/issues/20 for more context).

This was kind of a hack so far but worked fine until now. Because now we want to store a widget that is created by assigning a widget to another widget which happens in `connect_components`. But at this point the struct is already initialized which means that the new widget can't be part of the widgets struct any more which makes it hard to change the title of a ViewStackPage later on for example. 

TL;DR The current design of Relm4 would largely restrict the usability of the syntax we came up with.

As for a solution I've though about changing the order the widgets are created. The model should still be created from the root to the leaves, but I think maybe the widget can be created from the leaves to the root. With this, a parent component would already have access to its child component when creating its widgets so initialization and connecting the widgets can take place in one function which would solve the problem.

P. S. You don't need to bother about the technical details I wrote here, but I wrote this in case you @tronta or someone else is interested in this matter :)

**Comment by tronta** [comment]
Hi @AaronErhardt, I'm really thankful that you keep me updated at this level of detail. Hope it does not impact too much the previous concept of Relm4.

**Comment by AaronErhardt** [comment]
@tronta I've mostly finished my work on the rework-core branch. The new macro syntax already works now and only required a few changes to the core architecture. I just need to port relm4-components and several examples to the new version.

**Comment by tronta** [comment]
What is the right way to apply a command on an optional result?
``` rust  
                  add_titled(Some("First"), "First Page") = &gtk::Label {
                        set_label: "This is the start page",
                    } ->  {
                        set_icon_name?: Some("")
                    },
```
something like this?
``` rust  
                  add_titled(Some("First"), "First Page") = &gtk::Label {
                        set_label: "This is the start page",
                    }? ->  {
                        set_icon_name?: Some("")
                    },
```

**Comment by AaronErhardt** [comment]
@tronta I totally forgot that libadwaita returns `Option` there. I thought that optional values in this place wouldn't make sense because, well, how could this fail, right? So there is no syntax for this yet.

Even though it wouldn't be terribly complex to add such a syntax, I would rather suggest asking the maintainers of libadwaita-rs to make this non-optional because as far as I understand the C code this can not fail and shouldn't be optional.

**Comment by tronta** [comment]
I guess you are right. I can create a ticket and ask. I'm trying to think about cases when this makes sense…

**Comment by tronta** [comment]
https://gitlab.gnome.org/World/Rust/libadwaita-rs/-/issues/11

**Comment by AaronErhardt** [comment]
Mhh, seems like this won't be fixed upstream soon, but still I don't quite understand their reasoning. They seems to have discovered the same thing over 4 months ago and still this API persists...

**Comment by tronta** [comment]
Yes, indeed.

**Comment by AaronErhardt** [comment]
@tronta I've added the syntax to the macro now. It should look like `-> ? { ... }`. I'll shortly release 0.4.0-beta.2 for this.

**Comment by AaronErhardt** [comment]
0.4.0-beta.2 is out, let test it then :)

**Comment by tronta** [comment]
What do you mean? Have you already implemented it? You are fast. What is the exact syntax?

**Comment by AaronErhardt** [comment]
As I said, this wasn't really hard to add. I actually had the parsing part already but never implemented the code generation. 

So this is the syntax (question mark comes after the arrow):

```rust
                  add_titled(Some("First"), "First Page") = &gtk::Label {
                        set_label: "This is the start page",
                    } -> ? {
                        set_icon_name?: Some("")
                    },
```

Also, different from the normal optional assign this actually calls `unwrap()` and would panic. But that should be fine.

**Comment by tronta** [comment]
Currently it seams like one needs to provide an explicit type of the returned value eg
``` rust
                    add_titled(Some("First"), "First Page") = &gtk::Label {
                        set_label: "This is the start page",
                    } -> first_page:ViewStackPage? {
                        set_icon_name: Some("eye-open-negative-filled-symbolic"),
                        set_needs_attention: true
                    },
```
would it be also possible to not provide the type like this?
``` rust
                    add_titled(Some("First"), "First Page") = &gtk::Label {
                        set_label: "This is the start page",
                    } -> first_page? {
                        set_icon_name: Some("eye-open-negative-filled-symbolic"),
                        set_needs_attention: true
                    },
```
As the return type is usually pretty defined.

**Comment by AaronErhardt** [comment]
Yes, this is because the macro puts named returned widgets into the widgets struct. And there the code generation needs to generate a regular struct definition that requires a type. If it wasn't for the struct, Rust is smart enough to guess the type and therefore unnamed returned widgets that are not part of the struct don't need a type annotation.

```rust
Widgets {
    ...
    first_page: ViewStackPage,
}
```

**Comment by tronta** [comment]
Just to understand it. The macro expands to this:
``` rust
        let first_page: ViewStackPage = stack
            .add_titled(&_gtk_label_2, Some("First"), "First Page")
            .unwrap();
        stack.add_titled(&_gtk_label_3, Some("Second"), "Second Page");
        stack.add_titled(&_gtk_label_4, Some("Third"), "Third Page");
[…]
        _adw_viewswitchertitle_0.set_title("A stack switcher App");
        _adw_viewswitchertitle_0.set_stack(Some(&stack));
        _gtk_label_2.set_label("This is the start page");
```
It would work equally if the type would not be provided
``` rust
        let first_page = stack
            .add_titled(&_gtk_label_2, Some("First"), "First Page")
            .unwrap();
        stack.add_titled(&_gtk_label_3, Some("Second"), "Second Page");
        stack.add_titled(&_gtk_label_4, Some("Third"), "Third Page");
[…]
        _adw_viewswitchertitle_0.set_title("A stack switcher App");
        _adw_viewswitchertitle_0.set_stack(Some(&stack));
        _gtk_label_2.set_label("This is the start page");
```
So I'm not sure, what you exactly  meant.

**Comment by AaronErhardt** [comment]
Yes that's one part. But the type is actually only for the struct because the widget macro also defines the struct.

```rust
stuct AppWidgets {
    [...]
    first_page: ViewStackPage,
}

impl Widgets for AppWidgets {
[...]
    fn init_view() {
        let first_page: ViewStackPage = stack
            .add_titled(&_gtk_label_2, Some("First"), "First Page")
            .unwrap();
[...]
        first_page.set_icon_name(Some("eye-open-negative-filled-symbolic"));
        first_page.set_needs_attention(true);
[...]
        AppWidgets {
            [...]
            first_page,
        }
    }
}
```

**Comment by tronta** [comment]
ah, thanks a lot. I'm still too tired too think properly. I missed that part.
Cool work.

**Comment by tronta** [comment]
Still room for improvements, but maybe already good enough for merging,
What do you think?

**Comment by tronta** [comment]
I'd like still to add, that the `ViewSwitcherTitle` switches automatically to the bottom when the width is too narrow.
I'd assume this works via
``` xml
          <object class="AdwViewSwitcherBar" id="switcher_bar">
            <property name="stack">stack</property>
            <property name="reveal" bind-source="switcher_title" bind-property="title-visible" bind-flags="sync-create"/>
          </object>
```

**Comment by AaronErhardt** [comment]
The XML you copied here just sets the stack of the ViewSwitcherBar and binds the "reveal" property to the "title-visible" property of the title. This does only half of the job to make this work. What's done in the libadwaita demo is the following:

The title has two ViewSwitcherBars and one WindowTitle. The first, wide ViewSwitcherBar shows text next to the icons and the second, smaller one just the icons with a small text blow. And the WindowTitle just shows the title.

Then there is another (the third!) ViewSwitcherBar at the bottom. An that's where the property binding comes into play. All title widgets are placed in a Squeezer, that will show the largest of the three widgets that still fits on the screen. This causes the two ViewSwitcherBars to be hidden once the window is too small for them so the title is shown instead. And then the binding takes care of the ViewSwitcherBar at the bottom and shows is when the title is visible.

I hope that makes sense, I recommend running the adwaita demo with inspector to see this mechanism in action :)

**Comment by AaronErhardt** [comment]
Thanks a lot for your work! I think, too, that this can be merged now :)
Sorry for the delay...

**Comment by AaronErhardt** [comment]
@tronta here's the code for adding the bottom bar: https://github.com/AaronErhardt/relm4/commit/11cc880ef47b3f93cd8a9b623b08eb8de545b668
